api/actions/views.py

```diff
@@ -0,0 +1,191 @@
+        to_state                        string                              state of the reviewable after this action was created
+        comment                         string                              comment explaining the state change
+        trigger                         string                              name of the trigger for this action
+
+    ##Relationships
+
+    ###Target
+    Link to the object (e.g. preprint) this action acts on
+
+    ###Provider
+    Link to detail for the target object's provider
+
+    ###Creator
+    Link to the user that created this action
+
+    ##Links
+    - `self` -- Detail page for the current action
+
+    ##Query Params
+
+    + `page=<Int>` -- page number of results to view, default 1
+
+    + `filter[<fieldname>]=<Str>` -- fields and values to filter the search results on.
+
+    Actions may be filtered by their `id`, `from_state`, `to_state`, `date_created`, `date_modified`, `creator`, `provider`, `target`
+
+    ###Creating New Actions
+
+    Create a new Action by POSTing to `/actions/`, including the target preprint and the action trigger.
+
+    Valid triggers are: `submit`, `accept`, `reject`, and `edit_comment`
+
+        Method:        POST
+        URL:           /actions/
+        Query Params:  <none>
+        Body (JSON):   {
+                        "data": {
+                            "attributes": {
+                                "trigger": {trigger},           # required
+                                "comment": {comment},
+                            },
+                            "relationships": {
+                                "target": {                     # required
+                                    "data": {
+                                        "type": "preprints",
+                                        "id": {preprint_id}
+                                    }
+                                },
+                            }
+                        }
+                    }
+        Success:       201 CREATED + action representation
+    """
+    permission_classes = (
+        permissions.IsAuthenticatedOrReadOnly,
+        base_permissions.TokenHasScope,
+        reviews_permissions.ActionPermission,
+    )
+
+    required_read_scopes = [CoreScopes.NULL]
+    required_write_scopes = [CoreScopes.ACTIONS_WRITE]
+
+    parser_classes = (JSONAPIMultipleRelationshipsParser, JSONAPIMultipleRelationshipsParserForRegularJSON,)
+
+    serializer_class = ActionSerializer
+
+    view_category = 'actions'
+    view_name = 'create-action'
+
+    # overrides ListCreateAPIView
+    def perform_create(self, serializer):
+        target = serializer.validated_data['target']
+        self.check_object_permissions(self.request, target)
+
+        trigger = serializer.validated_data['trigger']
+        permission = reviews_permissions.TRIGGER_PERMISSIONS[trigger]
+        if permission is not None and not self.request.user.has_perm(permission, target.provider):
```

Comment:
Since it depends on the `trigger` value I put this check after the serializer had validated. Do you think it would be better to have some validation logic in `ActionPermission` (and raise 400 if need be), or to have some permission logic in the view? I guess the former seems cleaner...

Reply:
If possible, it's better to have permissions checks in the permissions classes for detail and action views. If not possible, or really messy, then do what you need to do.